### PR TITLE
Add --fake flag to migrate and rollback commands

### LINKED ISSUE #841: Is there a way to fake initial migration?
The idea is to create an initial migration and register it as applied in an existent database (phinxlog table). Here is an example [in Django migrations](https://docs.djangoproject.com/en/1.9/ref/django-admin/#cmdoption-migrate--fake-initial).

Is this possible with Phinx?

### PR DESCRIPTION
**Pull request by MasterOdin**
Sort of implements what #841 is asking fo with a different implementation than #266 (more inline with how Django's migration tool works). Though, this doesn't add a 'fake-initial' command, you'd just have to use fake with the appropriate options to just run it for your initial migration if you only wanted to fake that.

Both migrate and rollback accept a `--fake` flag that when used will mark any specified migrations/rollbacks as having been done, but won't actually run the up/down/change methods for the migration.

Usage:
```
bin/phinx migrate --help                                                                 [11:30:28]
Usage:
  migrate [options]

Options:
  -c, --configuration=CONFIGURATION  The configuration file to load
  -p, --parser=PARSER                Parser used to read the config file. Defaults to YAML
  -e, --environment=ENVIRONMENT      The target environment
  -t, --target=TARGET                The version number to migrate to
  -d, --date=DATE                    The date to migrate to
  -x, --dry-run                      Dump query to standard output instead of executing it
      --fake                         Mark any migrations selected as run, but don't actually execute them
  -h, --help                         Display this help message
  -q, --quiet                        Do not output any message
  -V, --version                      Display this application version
      --ansi                         Force ANSI output
      --no-ansi                      Disable ANSI output
  -n, --no-interaction               Do not ask any interactive question
  -v|vv|vvv, --verbose               Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
  The migrate command runs all available migrations, optionally up to a specific version

  phinx migrate -e development
  phinx migrate -e development -t 20110103081132
  phinx migrate -e development -d 20110103
  phinx migrate -e development -v

bin/phinx rollback --help                                                                [11:31:27]
Usage:
  rollback [options]

Options:
  -c, --configuration=CONFIGURATION  The configuration file to load
  -p, --parser=PARSER                Parser used to read the config file. Defaults to YAML
  -e, --environment=ENVIRONMENT      The target environment
  -t, --target=TARGET                The version number to rollback to
  -d, --date=DATE                    The date to rollback to
  -f, --force                        Force rollback to ignore breakpoints
  -x, --dry-run                      Dump query to standard output instead of executing it
      --fake                         Mark any rollbacks selected as run, but don\'t actually execute them
  -h, --help                         Display this help message
  -q, --quiet                        Do not output any message
  -V, --version                      Display this application version
      --ansi                         Force ANSI output
      --no-ansi                      Disable ANSI output
  -n, --no-interaction               Do not ask any interactive question
  -v|vv|vvv, --verbose               Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
  The rollback command reverts the last migration, or optionally up to a specific version

  phinx rollback -e development
  phinx rollback -e development -t 20111018185412
  phinx rollback -e development -d 20111018
  phinx rollback -e development -v
  phinx rollback -e development -t 20111018185412 -f

  If you have a breakpoint set, then you can rollback to target 0 and the rollbacks will stop at the breakpoint.
  phinx rollback -e development -t 0

  The version_order configuration option is used to determine the order of the migrations when rolling back.
  This can be used to allow the rolling back of the last executed migration instead of the last created one, or combined
  with the -d|--date option to rollback to a certain date using the migration start times to order them.
```

Test Migration:
```
<?php

use Phinx\Migration\AbstractMigration;

class Test extends AbstractMigration
{
    public function up()
    {
        $table = $this->table('test_table');
        $table->addColumn('name', 'string')->create();
        $this->insert('test_table', [['name' => 'test1'], ['name' => 'test2']]);
    }

    public function down()
    {
        $this->dropTable('test_table');
    }
}
```

Testing CLI:
```
root@8dc2af648658:/src# bin/phinx status
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
ordering by creation time

 Status  [Migration ID]  Started              Finished             Migration Name
----------------------------------------------------------------------------------
   down  20180302161813                                            Test

root@8dc2af648658:/src# mysql -Ddevelopment_db -e 'SELECT * FROM test_table;'
ERROR 1146 (42S02) at line 1: Table 'development_db.test_table' doesn't exist
root@8dc2af648658:/src# bin/phinx migrate
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db

 == 20180302161813 Test: migrating
 == 20180302161813 Test: migrated 0.0238s

All Done. Took 0.0847s
root@8dc2af648658:/src# bin/phinx status
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
ordering by creation time

 Status  [Migration ID]  Started              Finished             Migration Name
----------------------------------------------------------------------------------
     up  20180302161813  2018-03-02 16:28:46  2018-03-02 16:28:46  Test

root@8dc2af648658:/src# mysql -Ddevelopment_db -e 'SELECT * FROM test_table;'
+----+-------+
| id | name  |
+----+-------+
|  1 | test1 |
|  2 | test2 |
+----+-------+
root@8dc2af648658:/src# bin/phinx rollback
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db
ordering by creation time

 == 20180302161813 Test: reverting
 == 20180302161813 Test: reverted 0.0075s

All Done. Took 0.0692s
root@8dc2af648658:/src# bin/phinx status
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
ordering by creation time

 Status  [Migration ID]  Started              Finished             Migration Name
----------------------------------------------------------------------------------
   down  20180302161813                                            Test

root@8dc2af648658:/src# mysql -Ddevelopment_db -e 'SELECT * FROM test_table;'
ERROR 1146 (42S02) at line 1: Table 'development_db.test_table' doesn't exist
root@8dc2af648658:/src# bin/phinx migrate --fake
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db
warning performing fake migrations

 == 20180302161813 Test: migrating
 == 20180302161813 Test: migrated 0.0039s

All Done. Took 0.0720s
root@8dc2af648658:/src# bin/phinx status
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
ordering by creation time

 Status  [Migration ID]  Started              Finished             Migration Name
----------------------------------------------------------------------------------
     up  20180302161813  2018-03-02 16:29:06  2018-03-02 16:29:06  Test

root@8dc2af648658:/src# mysql -Ddevelopment_db -e 'SELECT * FROM test_table;'
ERROR 1146 (42S02) at line 1: Table 'development_db.test_table' doesn't exist
root@8dc2af648658:/src# bin/phinx rollback
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db
ordering by creation time

 == 20180302161813 Test: reverting

In PdoAdapter.php line 134:

  SQLSTATE[42S02]: Base table or view not found: 1051 Unknown table 'test_table'


rollback [-c|--configuration CONFIGURATION] [-p|--parser PARSER] [-e|--environment ENVIRONMENT] [-t|--target TARGET] [-d|--date DATE] [-f|--force] [-x|--dry-run] [--fake]

root@8dc2af648658:/src# bin/phinx rollback --fake
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db
ordering by creation time
warning performing fake rollbacks

 == 20180302161813 Test: reverting
 == 20180302161813 Test: reverted 0.0059s

All Done. Took 0.0718s
root@8dc2af648658:/src# bin/phinx status
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
ordering by creation time

 Status  [Migration ID]  Started              Finished             Migration Name
----------------------------------------------------------------------------------
   down  20180302161813                                            Test

root@8dc2af648658:/src# bin/phinx migrate
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db

 == 20180302161813 Test: migrating
 == 20180302161813 Test: migrated 0.0175s

All Done. Took 0.0689s
root@8dc2af648658:/src# bin/phinx rollback --fake
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db
ordering by creation time
warning performing fake rollbacks

 == 20180302161813 Test: reverting
 == 20180302161813 Test: reverted 0.0043s

All Done. Took 0.0759s
root@8dc2af648658:/src# mysql -Ddevelopment_db -e 'SELECT * FROM test_table;'
+----+-------+
| id | name  |
+----+-------+
|  1 | test1 |
|  2 | test2 |
+----+-------+
root@8dc2af648658:/src# bin/phinx migrate
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db

 == 20180302161813 Test: migrating

In PdoAdapter.php line 134:

  SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'test_table' already exists


migrate [-c|--configuration CONFIGURATION] [-p|--parser PARSER] [-e|--environment ENVIRONMENT] [-t|--target TARGET] [-d|--date DATE] [-x|--dry-run] [--fake]

root@8dc2af648658:/src# bin/phinx migrate --fake
Phinx by CakePHP - https://phinx.org. 0.9.2

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /src/db/migrations
using seed paths
warning no environment specified, defaulting to: development
using adapter mysql
using database development_db
warning performing fake migrations

 == 20180302161813 Test: migrating
 == 20180302161813 Test: migrated 0.0042s

All Done. Took 0.0991s
root@8dc2af648658:/src#
```

### CODE REVIEW

**Comment by dereuromark** [comment]
What is the differente of this vs `--mark-migrated ...`?

**Comment by MasterOdin** [comment]
I cannot find reference of a flag like that existing anywhere in Phinx?

**Comment by dereuromark** [comment]
I am sry, I was using it in cakephp/bake plugin (a wrapper around this library), which has a [command](https://github.com/cakephp/migrations/blob/master/src/Command/MarkMigrated.php) here for it.
But you are right, inside phinx there is no such thing yet.

**Comment by MasterOdin** [comment]
Alright cool, glad to have cleared that up so quickly. Please let me know if there's anything more you need from me for this PR.

File: src/Phinx/Console/PhinxApplication.php [hunk 1]
```diff
@@ -45,12 +45,14 @@ class PhinxApplication extends Application
      *
      * Initialize the Phinx console application.
      *
-     * @param string $version The Application Version
+     * @param string $version The Application Version, if null, use version out of composer.json file
```
Comment (dereuromark) [inline src/Phinx/Console/PhinxApplication.php]: string|null

Comment (dereuromark) [reply src/Phinx/Console/PhinxApplication.php]: Why is this part of the change anyway?

Comment (MasterOdin) [reply src/Phinx/Console/PhinxApplication.php]: Sorry, I had just added this separately and was going to put this in a separate PR as I'm a bit confused as to what was the full intention here as it used to be that the actual version string was passed to the function, but then it was changed to load this out of composer.json file, but all of the tests still want to pass in a variable when initializing the class.

I can revert the change to the file here and then just open an issue with the above to get an idea as to what the desired functionality would be.

Comment (dereuromark) [reply src/Phinx/Console/PhinxApplication.php]: I think you can fix up the doc block and that should be fine.

Comment (MasterOdin) [reply src/Phinx/Console/PhinxApplication.php]: Done

**Comment by lorenzo** [comment]
Thanks!

**Pull request merged**